File: server/functions/src/models/domain/assets/business_asset.ts
import { Asset } from '../asset';
import { Entity } from '../../../core/domain/entity';
import { Strings } from '../../../resources/strings';
import { Income } from '../income';

export interface BusinessAsset extends Asset {
  readonly buyPrice: number;
  readonly downPayment: number;
  readonly fairPrice: number;
  readonly passiveIncomePerMonth: number;
  readonly payback: number;
  readonly sellProbability: number;
}

export namespace BusinessAssetEntity {
  export const validate = (asset: any) => {
    const entity = Entity.createEntityValidator<BusinessAsset>(asset, 'Business Asset');

    entity.hasNumberValue('buyPrice');
    entity.hasNumberValue('downPayment');
    entity.hasNumberValue('fairPrice');
    entity.hasNumberValue('passiveIncomePerMonth');
    entity.hasNumberValue('payback');
    entity.hasNumberValue('sellProbability');

    entity.checkWithRules([
      [(a) => a.buyPrice <= 0, "BuyPrice can't be <= 0"],
      [(a) => a.downPayment <= 0, "DownPayment can't be <= 0"],
      [(a) => a.fairPrice <= 0, "FairPrice can't be <= 0"],
      [(a) => a.sellProbability <= 0, "SellProbability can't be <= 0"],
    ]);
  };

  export const getIncomeValue = (asset: BusinessAsset) => {
    return asset.passiveIncomePerMonth;
  };

  export const getIncome = (asset: BusinessAsset): Income => {
    return {
      id: asset.id || null,
      name: Strings.business(),
      value: getIncomeValue(asset),
      type: 'business',
    };
  };
}

File: client/lib/models/domain/game/possession_state/assets/business/business_asset.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../asset.dart';

part 'business_asset.freezed.dart';
part 'business_asset.g.dart';

@freezed
class BusinessAsset with _$BusinessAsset implements Asset {
  factory BusinessAsset({
    required String name,
    required AssetType type,
    required double buyPrice,
    required double downPayment,
    required double fairPrice,
    required double passiveIncomePerMonth,
    required double payback,
    required double sellProbability,
    required String id,
  }) = _BusinessAsset;

  factory BusinessAsset.fromJson(Map<String, dynamic> json) =>
      _$BusinessAssetFromJson(json);
}
